add the PERU_REEXEC_PYTHON heuristic on Windows

oconnor663 opened the pull request:
Normally we prefer to execute plugins directly. This is pretty reliable
on Unix, where scripts can specify their interpreter with a shebang
line.  However, it's not as reliable on Windows, because the
extension-interpreter mapping is a global config. In my experience,
installing and uninstalling a series of different Python interepreter
versions on a Windows machine can break that association, and as a
result break peru, in confusing ways. (Tests have also been broken by
default for this reason on every Windows CI provider I've ever tried.)

To work around this problem, we apply an extra heuristic on Windows: If
a plugin executable filename ends in .py, assume that we should
re-execute the current interpreter (sys.executable) to run that file,
rather than relying on the system shell to find the interpreter. This
fixes peru on systems with broken Python configs, and it makes no
difference in the vast majority of other cases.

For users who want to control this heuristic (either to disable it, or
to force the same behavior on Unix), we define the PERU_REEXEC_PYTHON
env var.

We discussed the previous behavior six years ago as part of
be35c14f9986f5481fc6c0e6aff04f08f401309c. At the time, I didn't want to
special-case Python plugins over any other language. Since then,
repeated problems on Windows have worn me down, particularly related to
CI. With the recent switch to GitHub Actions, I was looking at solving
this config issue yet again for a new platform, and I just decided I'd
rather have a permanent fix.

oconnor663 commented:
My first version of this PR forgot to actually check for the `.py` extension :p It passed tests because we suppress the rsync plugin on Windows. (It's a bash script.)

Any suggestions for tests that we could add here?

olson-sean-k commented:
> Any suggestions for tests that we could add here?

If we include a non-Python plugin for Windows, then CI would catch this. I suspect that we'd encounter the same issue though: whatever toolchain that plugin uses, it may not execute reliably across various Windows installations. Maybe a VBScript plugin could work. 😝 (Seriously though, VBScript seems to Just Work™ from the shell and desktop even on Windows 10.)

olson-sean-k commented on peru/plugin.py:
```diff
@@ -99,11 +148,11 @@ async def _plugin_job(plugin_context, module_type, module_fields, command, env,
                 cwd=plugin_context.cwd,
                 env=complete_env,
-                shell=True)
+                shell=is_shell_mode)
```
It looks like we previously always shelled out (enabled shell mode), but this change conditionally enables shell mode exclusively on Windows (when not using `sys.executable` with Python plugins). Is that okay? (Seems fine to me, but I wanted to highlight this since there may be some subtle behavioral changes when using `create_subprocess_exec`. :shrug:)

oconnor663 replied:
Yeah previously I guess I thought we might as well do it the same way everywhere for consistency. We could keep doing it everywhere now too. But I realized while I was refactoring this code, that `list2cmdline` was only ever intended to match the behavior of the Windows shell, and it's kind of a coincidence that it produces output that's been working for us on Unix, and there might be some cases where it breaks things that we don't know about. (Also the source module for that function includes the comment "NOTE: We intentionally exclude list2cmdline as it is considered an internal implementation detail.")

You're right though: In an internal-implementation-details sort of sense, it is a big change to how things work. It's possible that it breaks something surprising. But for the most part, if all our tests pass on Linux/Windows/macOS, I'm pretty confident we're ok here. And if I think about weird hypothetical breaks, it's a lot easier for me to come up with those hypotheticals in shell mode than it is for me to come up with them if we don't use the shell. So maybe this will be a win for less weird breakage overall?

olson-sean-k replied:
> ... it's a lot easier for me to come up with those hypotheticals in shell mode ... So maybe this will be a win for less weird breakage overall?

Yeah, this is probably the right way to do this. LGTM.

oconnor663 commented:
Yeah or maybe just a `.bat` file plugin. We already disable the tests that depend on rsync when we run tests on Windows, so it wouldn't be any trouble to write a new test that's disabled in the other direction. I'll add something.

oconnor663 commented:
Ok, I rebased the new test commit to come first, so you can test it both with and without the new heuristic. Both pass for me localle. And I fixed my mistake with `list2cmdline`, where I (ironically) forgot that its argument should be a list.

oconnor663 commented:
🚢
